libjolie/src/main/java/jolie/lang/parse/module/SymbolReferenceResolver.java

```diff
@@ -0,0 +1,782 @@
+				rrBranch.body.accept( this );
+			}
+		}
+
+		@Override
+		public void visit( NotificationForwardStatement n ) {}
+
+		@Override
+		public void visit( SolicitResponseForwardStatement n ) {}
+
+		@Override
+		public void visit( InterfaceExtenderDefinition n ) {
+			if( n.defaultOneWayOperation() != null ) {
+				n.defaultOneWayOperation().accept( this );
+			}
+			if( n.defaultRequestResponseOperation() != null ) {
+				n.defaultRequestResponseOperation().accept( this );
+			}
+			for( OperationDeclaration op : n.operationsMap().values() ) {
+				op.accept( this );
+			}
+		}
+
+		@Override
+		public void visit( InlineTreeExpressionNode n ) {
+			for( Operation operation : n.operations() ) {
+				if( operation instanceof InlineTreeExpressionNode.AssignmentOperation ) {
+					((InlineTreeExpressionNode.AssignmentOperation) operation).expression()
+						.accept( this );;
+				} else if( operation instanceof InlineTreeExpressionNode.DeepCopyOperation ) {
+					((InlineTreeExpressionNode.DeepCopyOperation) operation).expression()
+						.accept( this );;
+				}
+			}
+		}
+
+		@Override
+		public void visit( VoidExpressionNode n ) {}
+
+		@Override
+		public void visit( ProvideUntilStatement n ) {
+			n.provide().accept( this );
+			n.until().accept( this );
+		}
+
+		@Override
+		public void visit( TypeChoiceDefinition n ) {
+			n.left().accept( this );
+			if( n.right() != null ) {
+				n.right().accept( this );
+			}
+		}
+
+		@Override
+		public void visit( ImportStatement n ) {}
+
+		private Optional< SymbolInfo > getSymbol( ParsingContext context, String name ) {
+			Optional< SymbolInfo > symbol = Helpers.firstNonNull( () -> {
+				if( !symbolTables.containsKey( currentURI )
+					|| !symbolTables.get( currentURI ).symbol( name ).isPresent() ) {
+					return null;
+				}
+				return symbolTables.get( currentURI ).symbol( name ).get();
+			}, () -> {
+				if( !symbolTables.containsKey( context.source() )
+					|| !symbolTables.get( context.source() ).symbol( name ).isPresent() ) {
+					return null;
+				}
+				return symbolTables.get( context.source() ).symbol( name ).get();
+			} );
+			return symbol;
+		}
+	}
+
+	/**
+	 * perform lookup to external symbol's source of AST node
+	 * 
+	 * @throws SymbolNotFoundException
+	 */
+	private SymbolInfo symbolSourceLookup( SymbolInfoExternal symbolInfo )
+		throws SymbolNotFoundException {
+		ModuleRecord externalSourceRecord =
+			this.moduleMap.get( symbolInfo.moduleSource().get().source() );
+		Optional< SymbolInfo > externalSourceSymbol =
+			externalSourceRecord.symbol( symbolInfo.moduleSymbol() );
+		if( !externalSourceSymbol.isPresent() ) {
+			throw new SymbolNotFoundException( symbolInfo.name(), symbolInfo.moduleTargets() );
+		}
+		if( externalSourceSymbol.get().scope() == Scope.LOCAL ) {
+			return externalSourceSymbol.get();
+		} else {
+			return symbolSourceLookup( (SymbolInfoExternal) externalSourceSymbol.get() );
```

Comment:
Can't this loop infinitely?
You need a parameter `Set< URI >` to remember the modules that you have already visited.